Add eks addon resource, data_source

juozasget opened the pull request:
This is my first contribution - feedback is very much welcome.

Thank you!

*UPDATED & removed WIP mention* 

<!--- See what makes a good Pull Request at : https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #16542

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
data-source/aws_eks_addon: new data-source
resource/aws_eks_addon: new resource
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make testacc TESTARGS='-run=TestAccAWSEksAddon*'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEksAddon* -timeout 120m
=== RUN   TestAccAWSEksAddonDataSource_basic
=== PAUSE TestAccAWSEksAddonDataSource_basic
=== RUN   TestAccAWSEksAddon_basic
=== PAUSE TestAccAWSEksAddon_basic
=== RUN   TestAccAWSEksAddon_disappears
=== PAUSE TestAccAWSEksAddon_disappears
=== RUN   TestAccAWSEksAddon_disappears_Cluster
=== PAUSE TestAccAWSEksAddon_disappears_Cluster
=== RUN   TestAccAWSEksAddons_AddonVersion
=== PAUSE TestAccAWSEksAddons_AddonVersion
=== RUN   TestAccAWSEksAddons_ResolveConflicts
=== PAUSE TestAccAWSEksAddons_ResolveConflicts
=== RUN   TestAccAWSEksAddons_ServiceAccountRoleArn
=== PAUSE TestAccAWSEksAddons_ServiceAccountRoleArn
=== RUN   TestAccAWSEksAddons_Tags
=== PAUSE TestAccAWSEksAddons_Tags
=== CONT  TestAccAWSEksAddonDataSource_basic
=== CONT  TestAccAWSEksAddons_ResolveConflicts
=== CONT  TestAccAWSEksAddon_disappears_Cluster
=== CONT  TestAccAWSEksAddons_AddonVersion
=== CONT  TestAccAWSEksAddon_basic
=== CONT  TestAccAWSEksAddon_disappears
=== CONT  TestAccAWSEksAddons_ServiceAccountRoleArn
=== CONT  TestAccAWSEksAddons_Tags
--- PASS: TestAccAWSEksAddons_AddonVersion (848.97s)
--- PASS: TestAccAWSEksAddon_basic (889.61s)
--- PASS: TestAccAWSEksAddonDataSource_basic (938.24s)
--- PASS: TestAccAWSEksAddons_ServiceAccountRoleArn (977.66s)
--- PASS: TestAccAWSEksAddon_disappears (981.65s)
--- PASS: TestAccAWSEksAddon_disappears_Cluster (986.26s)
--- PASS: TestAccAWSEksAddons_ResolveConflicts (1118.40s)
--- PASS: TestAccAWSEksAddons_Tags (1412.16s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1413.873s
...
```

shuheiktgw commented on aws/resource_aws_eks_addon.go:
```diff
@@ -0,0 +1,332 @@
+			},
+			"resolve_conflicts": {
+				Type:     schema.TypeString,
+				Optional: true,
```
It would be nicer if we had a validation like:

```
ValidateFunc: validation.StringInSlice(eks.ResolveConflicts_Values(), false),
```

shuheiktgw commented on aws/resource_aws_eks_addon.go:
```diff
@@ -0,0 +1,332 @@
+				Required: true,
+				ForceNew: true,
+			},
+			"cluster_name": {
```
I guess we need `client_request_token` attribute?

bflad replied:
Typically a `ClientRequestToken` API parameter should be automatically set in the input type via `ClientRequestToken: aws.String(resource.UniqueId()),`

shuheiktgw commented on aws/resource_aws_eks_addon.go:
```diff
@@ -0,0 +1,332 @@
+	d.Set("addon_version", addon.AddonVersion)
+	d.Set("service_account_role_arn", addon.ServiceAccountRoleArn)
+	d.Set("created_at", aws.TimeValue(addon.CreatedAt).String())
+	d.Set("modified_at", aws.TimeValue(addon.ModifiedAt).String())
```
I think you need to define `status`, `created_at` and `modified_at` field to the schema.

juozasget replied:
Yes for sure. I was unsure if we actually need these attributes initially.
Looking at some other AWS resources (resource_aws_eks_cluster for example) I see that all of these definitely should be added.

juozasget commented:
@shuheiktgw @bflad Thank you very much for your initial comments.

I have updated original post with newly added ACC test outputs along with other missing things. I think the PR is now in good shape for a review. Things I'd like to point out:

- Issue #16542 is only for resource, and this PR adds a data_source also. Let me know if you think it's not needed at all or if it should be moved and not done in scope of this PR.
- There are some interesting interactions to consider when creating EKS Addon, when unmanaged version exists or when updating the addon and there are conflicts. Would definitely like some feedback on this part and if we should make some decision for the user and handle it or we should just throw an error and let the user handle it (current implementation).

I also left some TODO items, which would be nice to do/improve. I will try to work on them while waiting for feedback. 🔧
Thank you for waiting!

juozasget commented:
Hello. I've addressed the TODO things and improved the PR, mainly in these areas:
- Added waiter package and moved relevant code there.
- Fixed error handling according to newly released guidelines.
- Improved error handling in the area I mentioned when add-on creation/update fails with information/warning for user.

I have been marinating this for 2-3 weeks, because after these changes **TestAccAWSEksAddons_Tags** started failing. I've been pulling my hair out trying to find what could have caused this I believe I've finally found it.

I added aws_eks_addon to aws_eks_cluster deps:
`resource.AddTestSweepers("aws_eks_cluster", &resource.Sweeper{
		Name: "aws_eks_cluster",
		F:    testSweepEksClusters,
		Dependencies: []string{
			"aws_eks_fargate_profile",
			"aws_eks_fargate_node_group",
		},
	})`

This caused the eks add-on deletion api to be called, before eks cluster deletion. Previously it would just delete the eks cluster. I believe that there might be some weird bug that will cause eks add-on to get stuck in DELETING state if the eks add-on had it's TAG values modified. This only happens if tag values were modified/set.

> 1. Create EKS addon cia CLI with some tags.
> 2. Modify the tags/put new tags on the created EKS addon via CLI.
> 3. Delete the EKS addon.
> 
> Result: addon is stuck in DELETING state for 24+ hours. Example of 1 hours DELETION attached.

Opened question in AWS forum. Sadly I don't think there is much chance of getting an answer...
https://forums.aws.amazon.com/thread.jspa?threadID=334229&tstart=0

I've since removed it from sweeper dependencies and all test are passing. Not sure what to do about this.

NeckBeardPrince commented:
Any chance we can review this PR again? Please.

DonBower commented:
is this going to be merged?

williamcodes commented:
Can we get this merged in? It's impossible to configure a fully functioning EKS cluster with managed nodes using only terraform right now because you can't install the VPC CNI addon.

Who needs to code review this? @shuheiktgw @bflad

gdavison reviewed (approved):
Thanks for the PR, @juozasget. I've made some updates

Acceptance test results

--- PASS: TestAccAWSEksAddon_basic (772.00s)
--- PASS: TestAccAWSEksAddon_ServiceAccountRoleArn (830.59s)
--- PASS: TestAccAWSEksAddonDataSource_basic (857.39s)
--- PASS: TestAccAWSEksAddon_Tags (885.01s)
--- PASS: TestAccAWSEksAddon_disappears_Cluster (891.55s)
--- PASS: TestAccAWSEksAddon_disappears (946.00s)
--- PASS: TestAccAWSEksAddon_ResolveConflicts (976.91s)
--- PASS: TestAccAWSEksAddon_AddonVersion (989.17s)
